ts/score/peerScore.ts

```diff
@@ -0,0 +1,596 @@
+   */
+  _markInvalidMessageDelivery (id: PeerId, message: Message): void {
+    const pstats = this.peerStats.get(id)
+    if (!pstats) {
+      return
+    }
+
+    message.topicIDs.forEach(topic => {
+      const tstats = ensureTopicStats(topic, pstats, this.params)
+      if (!tstats) {
+        return
+      }
+
+      tstats.invalidMessageDeliveries += 1
+    })
+  }
+
+  /**
+   * Increments the "first message deliveries" counter for all scored topics the message is published in,
+   * as well as the "mesh message deliveries" counter, if the peer is in the mesh for the topic.
+   * @param {PeerId} id
+   * @param {Message} message
+   * @returns {void}
+   */
+  _markFirstMessageDelivery (id: PeerId, message: Message): void {
+    const pstats = this.peerStats.get(id)
+    if (!pstats) {
+      return
+    }
+
+    message.topicIDs.forEach(topic => {
+      const tstats = ensureTopicStats(topic, pstats, this.params)
+      if (!tstats) {
+        return
+      }
+
+      let cap = this.params.topics[topic].firstMessageDeliveriesCap
+      tstats.firstMessageDeliveries += 1
+      if (tstats.firstMessageDeliveries > cap) {
+        tstats.firstMessageDeliveries = cap
+      }
+
+      if (!tstats.inMesh) {
+        return
+      }
+
+      cap = this.params.topics[topic].meshMessageDeliveriesCap
+      tstats.meshMessageDeliveries += 1
+      if (tstats.meshMessageDeliveries > cap) {
+        tstats.meshMessageDeliveries = cap
+      }
+    })
+  }
+
+  /**
+   * Increments the "mesh message deliveries" counter for messages we've seen before,
+   * as long the message was received within the P3 window.
+   * @param {PeerId} id
+   * @param {Message} message
+   * @param {number} validatedTime
+   * @returns {void}
+   */
+  _markDuplicateMessageDelivery (id: PeerId, message: Message, validatedTime = 0): void {
+    const pstats = this.peerStats.get(id)
+    if (!pstats) {
+      return
+    }
+
+    const now = validatedTime ? Date.now() : 0
+
+    message.topicIDs.forEach(topic => {
+      const tstats = ensureTopicStats(topic, pstats, this.params)
+      if (!tstats) {
+        return
+      }
+
+      if (!tstats.inMesh) {
+        return
+      }
+
+      const tparams = this.params.topics[topic]
+
+      // check against the mesh delivery window -- if the validated time is passed as 0, then
+      // the message was received before we finished validation and thus falls within the mesh
+      // delivery window.
+      if (validatedTime && now > validatedTime + tparams.meshMessageDeliveriesWindow) {
+        return
+      }
+
+      const cap = tparams.meshMessageDeliveriesCap
+      tstats.meshMessageDeliveries += 1
+      if (tstats.meshMessageDeliveries > cap) {
+        tstats.meshMessageDeliveries = cap
+      }
+    })
+  }
+
+  /**
+   * Gets the current IPs for a peer.
+   * @param {PeerId} id
+   * @returns {Array<string>}
+   */
+  _getIPs (id: PeerId): string[] {
+    return this._addressBook.getMultiaddrsForPeer(id)
```

Comment:
We should use here connectionManager. We might have more than one connection for a peer in some scenarios. So, maybe you should go to the internal data of the connectionManager instead of using `connectionManager.get` directly (you can see its implementation)

Reply:
gotcha, something like `get` but where the last `find` is a `filter` instead.

then here eg: `connectionManager.getOpenConnections(id).filter(c => c.remoteAddr)`

Reply:
I think that a `getOpenConnections` would be useful yes! Do you want to PR it and ask feedback to jacob?